ebos/ecloutputblackoilmodule.hh

```diff
@@ -1044,170 +1089,277 @@ public:
-                    pressureUnitConvert_(regHydroCarbonPoreVolumeAveragedPressure);
-                    outputRegionFluidInPlace_(tmpO, tmp, regHydroCarbonPoreVolumeAveragedPressure, reg + 1);
-                }
+    void outputFipLogImpl(const RegionSum& rsum) const {
+
+
+        {
+            Scalar fieldHydroCarbonPoreVolumeAveragedPressure = pressureAverage_(rsum.fieldPressurePvHydrocarbon,
+                                                                                 rsum.fieldPvHydrocarbon,
+                                                                                 rsum.fieldPressurePv,
+                                                                                 rsum.fieldFipValues[FipDataType::PoreVolume],
+                                                                                 true);
+            std::array<Scalar, FipDataType::numFipTypes> initial_values = *this->fieldInitialInplace_;
+            std::array<Scalar, FipDataType::numFipTypes> current_values = rsum.fieldFipValues;
+            fipUnitConvert_(initial_values);
+            fipUnitConvert_(current_values);
+
+            pressureUnitConvert_(fieldHydroCarbonPoreVolumeAveragedPressure);
+            outputRegionFluidInPlace_(initial_values,
+                                      current_values,
+                                      fieldHydroCarbonPoreVolumeAveragedPressure);
+        }
+
+        for (size_t reg = 0; reg < rsum.ntFip; ++reg) {
+            std::array<Scalar, FipDataType::numFipTypes> initial_values;
+            std::array<Scalar, FipDataType::numFipTypes> current_values;
+
+            for (std::size_t fip_type = 0; fip_type < FipDataType::numFipTypes; fip_type++) {
+                initial_values[fip_type] = this->regionInitialInplace_.value()[fip_type][reg];
+                current_values[fip_type] = rsum.regFipValues[fip_type][reg];
             }
+            fipUnitConvert_(initial_values);
+            fipUnitConvert_(current_values);
+
+            Scalar regHydroCarbonPoreVolumeAveragedPressure
+                = pressureAverage_(rsum.regPressurePvHydrocarbon[reg],
+                                   rsum.regPvHydrocarbon[reg],
+                                   rsum.regPressurePv[reg],
+                                   rsum.regFipValues[FipDataType::PoreVolume][reg],
+                                   true);
+            pressureUnitConvert_(regHydroCarbonPoreVolumeAveragedPressure);
+            outputRegionFluidInPlace_(initial_values, current_values, regHydroCarbonPoreVolumeAveragedPressure, reg + 1);
         }
+    }
+
+
 
+
+    // write Fluid In Place to output log
+    void outputFipLog(std::map<std::string, double>& miscSummaryData,  std::map<std::string, std::vector<double>>& regionData, const bool substep)
+    {
+        auto rsum_map = this->makeRegionSums();
+        if (!isIORank_())
+            return;
```

Comment:
I'll be honest&mdash;I think this code looks weird.  As a reader I wonder why we need to `makeRegionSums()` **before** we check if we're the IO rank when we only use the result on the IO rank.

First: Do we need to do that?  Second, if we do need to `makeRegionSums` on all processes, is there no other way of phrasing this?  I'd prefer if the function were named something along the lines of
```C++
accumulateAndExchangeRegionSums()
```

Reply:
Yes - this is weird; I have only refactored the existing code and not changed semantics / public API at all. The weirdness might stand more out now.

The `makeRegionSums()` accumulates over all processes - so it must be called globally, and then the results is only used on io rank. I'll rename the function to `accumulateRegionsSums()`